### PR TITLE
[bug 1130765] Implement Alerts API

### PR DESCRIPTION
**Pull request by willkg**
* add api_auth Token system for Django user agnostic API permission
  granting
* add alerts GET/POST API, docs and tests

r?

### CODE REVIEW

**Comment by willkg** [comment]
This is kind of a big PR. Sorry about that.

The only users for this will be Cheng and Robert. If there are bugs, it's not a big deal--we can fix them as we go along.

The things I'd love to get eyes on is the authorization/permission stuff and a review for "bad code".

File: fjord/alerts/admin.py [hunk 1]
```diff
@@ -5,7 +5,7 @@
 
 class AlertFlavorAdmin(admin.ModelAdmin):
     list_display = (
-        'enabled', 'name', 'slug', 'description', 'more_info',
+        'name', 'slug', 'enabled', 'description', 'more_info',
```
Comment (willkg) [inline fjord/alerts/admin.py]: Changing this means that "name" is the left-most column in the admin table and is the one that you click on for details. That's easier to use than "enabled".

File: fjord/urls.py [hunk 2]
```diff
@@ -8,11 +8,12 @@
 from fjord.base.monkeypatches import patch
 patch()
 
-from django.contrib import admin
+from django.contrib import admin  # noqa
```
Comment (willkg) [inline fjord/urls.py]: Had to add the `noqa` since we don't use it and flake8 got whiny.

File: docs/alerts_api.rst [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+
++-------------------+-------+--------------------------------------------------------+
+|field              |type   |notes                                                   |
++===================+=======+========================================================+
+|severity           |integer|This is the severity of the alert. 0 = don't care. 10 = |
+|                   |       |call the president.                                     |
++-------------------+-------+--------------------------------------------------------+
+|summary            |string |Brief summary of what the alert is about.               |
++-------------------+-------+--------------------------------------------------------+
+|description        |string |Involved description fo all the details that help you   |
```
Comment (mythmon) [inline docs/alerts_api.rst]: Typo. "fo" -> "of"

File: docs/alerts_api.rst [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+|                   |       |                                                        |
+|                   |       |You can make this a JSON encoded field if you have lots |
+|                   |       |of key/value pairs you want to include.                 |
++-------------------+-------+--------------------------------------------------------+
+|emitter_name       |string |Max length: 100.                                        |
+|                   |       |                                                        |
+|                   |       |The name of the emitter that created this alert. It     |
+|                   |       |could be a script name. It could be a sekret code-name  |
+|                   |       |for the emitter. It could be skynet. So long as we can  |
+|                   |       |distinguish it from other emitters, it's all good.      |
```
Comment (mythmon) [inline docs/alerts_api.rst]: I think you missed a good opportunity for a skynet joke here. Oh well.

**Comment by mythmon** [comment]
The auth and permissions look good. I'm glad that turned out to be straight forward. I assume the only way to create new flavors and tokens (and link the two) is you doing it in the admin?

Is there a reason you made the API views yourself instead of using the model mixins like [here](https://github.com/mozilla/kitsune/blob/master/kitsune/notifications/api.py#L66-L80)? I think it would simplify those a lot, though the thing you do with Links might not be doable, which would be a reason to do what you did instead of using the mixins. That would require more research, I guess.

This seems solid, and I think it achieve the goals we talked about for auth. I didn't run the code or anything, I only looked at the diffs, but Travis liked it.

**Comment by willkg** [comment]
I never got around to becoming enough of an expert in DRF to use the mixins effectively without a lot of struggling. I used them in richard (https://github.com/pyvideo/richard/blob/master/src/richard/videos/views.py#L316), but I find it a lot easier and maintainable to just go with subclassing APIView and doing my own thing. I have similar issues with Django CBVs. DRF 3 makes some API changes. It's possible that'll make it easier.

In this case, I'm not sure it helps a whole lot. Flavor comes from the url. Links are slightly complicated (though I've done similar stuff with richard).

I'm on the fence. I could rewrite it with generics. Do you think it'd be easier to augment and maintain?

Speaking of flavor, I'm not sure what I did is such a great idea. Maybe I should change it to `POST /api/v1/alerts/` and `GET /api/v1/alerts/<flavor>/`?

**Comment by mythmon** [comment]
One thing that could be better in that Richard view would be moving the filtering to a `filter_backend`, like [in the SUMO Question's API](https://github.com/mozilla/kitsune/blob/master/kitsune/questions/api.py#L132-L211). The filter backends tend to be more declarative, which I find helps with keeping things simple. Also, I'm pretty sure `get_queryset` isn't meant to do any filtering anyways, you would do that in something like `filter_queryset`. This is partially because `get_queryset` is used as the base for detail views too, not just list views.

About the url structure: Does it make sense to get a list of alerts without filtering on flavor? Like `GET /api/v1/alerts/`? If so, I'd actually make the api `GET /api/v1/alerts/?flavor=<flavor>` instead. At least, that is what I've been doing on on SUMO's v2 api. The general pattern I've been using is `/api/2/<model>/[id]/[action]/[?filter=value]`. The id and action don't apply to all queries, so sometimes that is `/api/2/user/generate` (which is the generate action on the user "model") or `/api/2/question/1234/` (which is the detail view for the 1234 question). Filters are always query parameters, and don't really apply to detail views.

> NB: Above the term "model" doesn't map 1:1 with Django's models, mainly because the "user" API model is actually a mix of `django.contrib.auth.models.User` and `kitsune.users.models.Profile`.

As far as switching to generics... I'm not sure. It looks to me like a lot of what you are doing is error checking and input validation, which the generics are really good with. On the other hand, I think that it would be hard to convince DRF 2.x to do what you want with the links. It also isn't really a single model API, since you tie a lot of behavior into the Flavor class stuff (which makes the generics less useful). I'd say if you want a non-trivial generics view to learn with, this would be that. In the end, maintainability is about who is writing the code, so if this is easier for you to maintain, I'd say stick with it. I don't think what you have is bad. I like CBV and the DRF generics mainly because they tend to make smaller functions (though more of them) instead of big monolithic view functions.

**Comment by willkg** [comment]
That makes sense. I think these view functions are small enough I'll just stick with what I've got. They're short and easy to reason about. I do think I'll rework the API urls. That'll reduce some of the complexity.

 I'll look into the richard stuff next time I'm fiddling with that API.

Thanks!

File: docs/alerts_api.rst [hunk 5]
```diff
@@ -0,0 +1,275 @@
...
+
+Alert flavors
+=============
+
+The alert flavor is the type of alert this is. It allows us to create
+a loose grouping of like alerts.
+
+For example, the "mfbt" alert flavor covers all alerts about mfbt.
+
+To get a token, `write up a bug
```
Comment (rlr) [inline docs/alerts_api.rst]: `To get a token,` doesn't make sense here. I think it's a copy/pasta mistake?

File: docs/alerts_api.rst (hunk 4 above)
Comment (willkg) [inline docs/alerts_api.rst]: I was super tired writing this and dealing with emacs table-mode. I'm totally game for anything you have in mind.

**Comment by willkg** [comment]
I factored the comments in (and fixed the docs typos) and rewrote the API endpoints like this:

POST `/api/v1/alerts/flavor/<flavor>/addalert/`

> POST pulls the flavor from the url, but everything else from the HTTP request body. There are good tests for allt he bits.

GET `/api/v1/alerts/alert/`

> GET can take two querystring arguments: flavors and max. It verifies each flavor before doing anything. There are good tests for all the bits.

**Comment by willkg** [comment]
^^^ That fixes the documentation examples.

**Comment by willkg** [comment]
This is ready for another r? round.

**Comment by mythmon** [comment]
It seems odd to use an action on the `/alerts/flavor` endpoint to create an alert, instead of using `POST /api/v1/alerts/alert {"flavor": "mfbt", ...}`. Flavor is a property of alerts, not the other way around, right?

**Comment by willkg** [comment]
Mmm... Flavor is a property of alerts, yes, but there's a permission component here so I think the better way to talk about this is less "I'm creating an alert" and more "For this flavor, I'm adding an alert" where the operation is on the flavor's alerts, not alerts in general. If there wasn't an authorization component on the flavor, I'd probably just have it be alerts in general.

**Comment by mythmon** [comment]
That doesn't seem like a good enough reason to have a surprising URL for creating alerts, or for having  asymmetry between creating and viewing alerts. Even the English translation ("for this flavor, I'm adding an alert") sounds awkward. I don't think it is any easier to code either way, but I do like consistent URL structures, and this doesn't feel consistent to me.

**Comment by willkg** [comment]
^^^ Changed the api endpoint to what you suggested. IT WILL BE BETTER!

**Comment by mythmon** [comment]
I've got nothing else to talk about. This looks good, well tested, and well documented. I say :boat:.

**Comment by willkg** [comment]
e712384 [bug 1130765] Implement Alerts API

Thank you both for working through this with me!

**Pull request merged**